File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/util/MapPack.java
package net.cyclestreets.util;

import java.io.File;
import java.io.FileInputStream;
import java.io.FilenameFilter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Environment;

public class MapPack
{
  private static String MAPSFORGE_VERSION = "0.9.1";

  public static void searchGooglePlay(final Context context) {
    final Intent play = new Intent(Intent.ACTION_VIEW);
    play.setData(Uri.parse("market://search?q=net.cyclestreets"));
    context.startActivity(play);
  }

  public static List<MapPack> availableMapPacks() {
    final List<MapPack> packs = new ArrayList<>();

    final File obbDir = new File(Environment.getExternalStorageDirectory(), "Android/obb");
    if (!obbDir.exists())
      return packs;

    for (final File mapDir : obbDir.listFiles(new CycleStreetsMapFilter())) {
      final File map = findMapFile(mapDir, "main.");
      final Properties props = mapProperties(mapDir);
      final String name = props.getProperty("title");
      final String version = props.getProperty("version");
      if (map == null || name == null)
        continue;

      packs.add(new MapPack(name, version, map));
    }

    return packs;
  }

  public static MapPack findByPackage(final String packageName) {
    for (final MapPack pack : availableMapPacks())
      if (pack.path().contains(packageName))
        return pack;
    return null;
  }

  private static File findMapFile(final File mapDir, final String prefix) {
    for (final File c : mapDir.listFiles())
      if (c.getName().startsWith(prefix))
        return c;
    return null;
  }

  private static Properties mapProperties(final File mapDir) {
    final Properties details = new Properties();
    try {
      final File detailsFile = findMapFile(mapDir, "patch.");
      details.load(new FileInputStream(detailsFile));
    }
    catch (IOException | RuntimeException e) {
    }
    return details;
  }

  private static class CycleStreetsMapFilter implements FilenameFilter  {
    public boolean accept(final File dir, final String name) {
      return name.contains("net.cyclestreets.maps");
    }
  }

  //////////////////////////////////////////////////////
  private final String name_;
  private final String path_;
  private final String version_;

  private MapPack(final String n,
            final String v,
            final File p) {
    name_ = n;
    path_ = p.getAbsolutePath();
    version_ = v;
  }

  public String name() { return name_; }
  public String path() { return path_; }
  public boolean current() { return MAPSFORGE_VERSION.equals(version_); }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/util/MapPack.kt
package net.cyclestreets.util

import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import net.cyclestreets.api.Maps
import net.cyclestreets.api.VectorMap
import java.io.File
import java.io.FileInputStream
import java.io.FilenameFilter
import java.io.IOException
import java.util.*

class MapPack private constructor(
        private val vectorMap: VectorMap,
        context: Context
) {
    val id get() = vectorMap.id
    val title get() = "${vectorMap.name}${if(downloaded) "" else " (Needs download)"}"
    val path = File(context.getExternalFilesDir(null), "${vectorMap.id}.map").absolutePath
    val downloaded get() = File(path).exists()

    fun download(context: Context) {
        val dm = context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
        val request = DownloadManager.Request(
                Uri.parse(vectorMap.url)
        )
        request.setTitle("${vectorMap.name} Map Pack")
        request.setAllowedNetworkTypes(DownloadManager.Request.NETWORK_WIFI or DownloadManager.Request.NETWORK_MOBILE)
        request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE)
        request.setDestinationInExternalFilesDir(context, null, "${vectorMap.id}.map")

        dm.enqueue(request)
    } // download

    companion object {
        fun availableMapPacks(context: Context): List<MapPack> {
            val maps = Maps.get() ?: return emptyList()

            return maps.map { m -> MapPack(m, context) }
        }

        @JvmStatic
        fun findById(context: Context, packId: String): MapPack? {
            return availableMapPacks(context).find { it.id == packId }
        }
    }
}